File: 08-number-guesser/num_guess.erl
%% Number guessing game written in Erlang
-module(num_guess).
-author("M.V.Harish Kumar").
-compile([{date, "2023-12-08"}]).
-export([main/0]).


game() -> game(rand:uniform(100), 1, []).
game(Random_Num, Guesses, Guessed_Nums) ->
    {ok, [User_Num]} = io:fread("Enter your guess: ", "~d"),
    if
        User_Num == Random_Num ->
            io:format("Yay! You've guessed the number in ~w chances~n", [Guesses]);

        (User_Num > 100) orelse (User_Num < 1) ->
            io:format("Guessed number is out of range~n"),
            game(Random_Num, Guesses+1, Guessed_Nums);

        User_Num < Random_Num ->
            io:format("You've guessed low~n"),
            case lists:member(User_Num, Guessed_Nums) of
                true -> game(Random_Num, Guesses, Guessed_Nums);
                false -> game(Random_Num, Guesses+1, [User_Num|Guessed_Nums])
            end;

        User_Num > Random_Num ->
            io:format("You've guessed high~n"),
            case lists:member(User_Num, Guessed_Nums) of
                true -> game(Random_Num, Guesses, Guessed_Nums);
                false -> game(Random_Num, Guesses+1, [User_Num|Guessed_Nums])
            end
    end.

main() ->
    io:format("Number Guessing Game~n"),
    io:format("Guess the number which the computer thinks of~n"),
    io:format("The number is between 1 to 100~n"),
    game(),
    erlang:halt(0).

File: 08-number-guesser/num_guess.f90
! Number guesser game written in Fortran
! Written by M.V.Harish Kumar on 14-05-2024
program num_guess
    implicit none
    real :: rand
    integer :: rnum, unum, guesses, len
    integer, allocatable :: guessed_nums(:)

    print '(a)', "Number Guessing Game"
    print '(a)', "Guess the number which the computer thinks of"
    print '(a)', "The number is between 1 to 100"
    
    ! Initialize random number generator
    call random_seed()
    call random_number(rand)
    rnum = 1 + floor(rand * 100)
    
    guesses = 0
    allocate(guessed_nums(0))

    do
        print '(a$)', "Enter your guess: "
        read(*,*) unum

        if (unum == rnum) then
            guesses = guesses + 1
            print '(a,i0,a)', "Yay! You've guessed the number in ", guesses, " guesses"
            exit
        else if (unum > 100 .or. unum < 1) then
            print '(a)', "Guessed number is out of range"
            guesses = guesses + 1
        else if (unum < rnum) then
            print '(a)', "You've guessed low"
            if (.not. any(guessed_nums == unum)) guesses = guesses + 1
        else
            print '(a)', "You've guessed high"
            if (.not. any(guessed_nums == unum)) guesses = guesses + 1
        end if
        len = size(guessed_nums)
        deallocate(guessed_nums)
        allocate(guessed_nums(len+1))
        guessed_nums(len+1) = unum
    end do

    deallocate(guessed_nums)
end program num_guess
